File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/BruteForce.kt
package sudokusolver.kotlin

sealed class BruteForceSolution
object NoSolutions : BruteForceSolution()
object MultipleSolutions : BruteForceSolution()
data class SingleSolution(val solution: Board<SudokuNumber>) : BruteForceSolution()

fun bruteForce(board: Board<SudokuNumber?>): BruteForceSolution {
    val trialAndError = board.toMutableBoard()

    fun bruteForce(rowIndex: Int, columnIndex: Int): BruteForceSolution {
        fun moveToNextCell(): BruteForceSolution {
            return if (columnIndex + 1 >= UNIT_SIZE) {
                bruteForce(rowIndex + 1, 0)
            } else {
                bruteForce(rowIndex, columnIndex + 1)
            }
        }

        return when {
            rowIndex >= UNIT_SIZE -> SingleSolution(trialAndError.mapCellsToBoard { cell -> cell!! })
            trialAndError[rowIndex, columnIndex] != null -> moveToNextCell()

            else -> {
                val rowInvalid = trialAndError.getRow(rowIndex).filterNotNull()
                val columnInvalid = trialAndError.getColumn(columnIndex).filterNotNull()
                val blockInvalid = trialAndError.getBlock(BlockIndex(rowIndex, columnIndex)).filterNotNull()
                val valid = SudokuNumber.values().toSet() - rowInvalid - columnInvalid - blockInvalid
                var singleSolution: SingleSolution? = null
                valid.forEach { guess ->
                    trialAndError[rowIndex, columnIndex] = guess
                    val intermediateSolution = moveToNextCell()
                    if (intermediateSolution == MultipleSolutions) {
                        return MultipleSolutions
                    } else if (intermediateSolution is SingleSolution) {
                        if (singleSolution == null) {
                            singleSolution = intermediateSolution
                        } else {
                            return MultipleSolutions
                        }
                    }
                }
                trialAndError[rowIndex, columnIndex] = null
                singleSolution ?: NoSolutions
            }
        }
    }

    return bruteForce(0, 0)
}

File: SudokuSolver_Swift/SudokuSolver_Swift/BruteForce.swift
private enum BruteForceError: Error {
    case noSolutions, multipleSolutions
}

func bruteForce(board: Board<SudokuNumber?>) throws -> Board<SudokuNumber> {
    if !board.cells.contains(nil) {
        let filledBoard = board.mapCells { $0! }
        if isSolved(board: filledBoard) {
            return filledBoard
        } else {
            throw BruteForceError.noSolutions
        }
    }
    var trialAndError = board
    
    func bruteForce(rowIndex: Int, columnIndex: Int) throws -> Board<SudokuNumber> {
        func moveToNextCell() throws -> Board<SudokuNumber> {
            if columnIndex + 1 >= unitSize {
                try bruteForce(rowIndex: rowIndex + 1, columnIndex: 0)
            } else {
                try bruteForce(rowIndex: rowIndex, columnIndex: columnIndex + 1)
            }
        }
        
        if rowIndex >= unitSize {
            return trialAndError.mapCells { cell in cell! }
        } else if trialAndError[rowIndex, columnIndex] != nil {
            return try moveToNextCell()
        } else {
            let rowInvalid = Set(trialAndError.getRow(rowIndex: rowIndex).compactMap { $0 })
            let columnInvalid = Set(trialAndError.getColumn(columnIndex: columnIndex).compactMap { $0 })
            let blockInvalid = Set(trialAndError
                .getBlock(blockIndex: getBlockIndex(rowIndex: rowIndex, columnIndex: columnIndex))
                .compactMap { $0 })
            let invalid = rowInvalid.union(columnInvalid).union(blockInvalid)
            let valid = Set(SudokuNumber.allCases).subtracting(invalid)
            var singleSolution: Board<SudokuNumber>?
            for guess in valid {
                trialAndError[rowIndex, columnIndex] = guess
                do {
                    let intermediateSolution = try moveToNextCell()
                    if singleSolution == nil {
                        singleSolution = intermediateSolution
                    } else {
                        throw BruteForceError.multipleSolutions
                    }
                } catch BruteForceError.noSolutions {
                    //Ignore.
                }
            }
            trialAndError[rowIndex, columnIndex] = nil
            if let singleSolution {
                return singleSolution
            } else {
                throw BruteForceError.noSolutions
            }
        }
    }
    
    return try bruteForce(rowIndex: 0, columnIndex: 0)
}

private func isSolved(board: Board<SudokuNumber>) -> Bool {
    fatalError("Implement Me!")
}
